### PR TITLE
Run tests in clean virtual environment

### PR DESCRIPTION
**Pull request by astrofrog**
This fixes https://github.com/joerick/cibuildwheel/issues/121 by making it so that once the wheels are built, the tests are run in an isolated environment. This ensures isolation from the build environment and guarantees that the package is correctly defining its build and test-time dependencies.

I originally attempted to solve this with direct calls to virtualenv, but ran into issues with some of the CI tests here, which suggested to me that this was not a robust solution. I therefore have now taken the approach of using tox. The process is as follows:

* Create an empty temporary directory
* Write a simple tox configuration to that directory that installs the wheels, the test dependencies, and runs the test command
* Run tox in that directory

The tox configuration whitelists all external commands since it's hard to know what non-Python commands people might be relying on. For now we also do not pass any environment variables into the test environment.

If there are concerns that this could break existing projects (it's always hard to account for all corner cases) this behavior could be made opt-in, with the previous behavior being the default. Another idea would be to make it opt-in initially, and then choose to make it the default later once it's been stress tested by some projects.

### CODE REVIEW

**Comment by astrofrog** [comment]
There are some CI failures here related to the ``cp27-cp27mu`` Linux wheels, investigating.

**Comment by astrofrog** [comment]
I've managed to fix most CI issues but there's still one failure on Travis, which I'll investigate shortly.

**Comment by astrofrog** [comment]
This is turning out to be more complicated than I thought it would be, so I'll need to put this on pause for a while!

File: cibuildwheel/linux.py [hunk 1]
```diff
@@ -85,40 +89,48 @@ def build(project_dir, output_dir, test_command, test_requires, before_build, bu
...
+            envlist = test
+            skipsdist = true
+
+            [testenv]
+            whitelist_externals = *
+            commands =
+                pip install $delocated_wheel
+                {test_requires_cmd}
+                {test_cmd_prepared}
+            EOL
```
Comment (joerick) [inline cibuildwheel/linux.py]: I'm not familiar with tox, so bear with me - what is tox doing for us here? I was expecting to see some virtualenv commands, but I guess tox is handling that on our behalf using these options. Could it perhaps be clearer by doing running the virtualenv commands directly? 

(In my head it's just a `virtualenv $test_dir/env`, then a subshell that calls `source $test_dir/env/bin/activate` before the package-specific stuff.... that feels clearer to me, and requires less knowledge)

Comment (joerick) [reply cibuildwheel/linux.py]: Apologies, I now see your note in the PR. I can't see the previous attempts using direct virtualenv calls in this PR, I'd be curious what issues you ran into.

Comment (astrofrog) [reply cibuildwheel/linux.py]: @joerick - I still have a backup of the virtualenv version locally so can push it to a branch for you to look at shortly. Essentially the issue is that sometimes when the virtualenv was set up, ``python -m pip`` would still point to the system install even though ``python`` pointed to the virtualenv version, and I couldn't find a way around it.

Comment (joerick) [reply cibuildwheel/linux.py]: @astrofrog please do - I might be able to help you on that!

Comment (astrofrog) [reply cibuildwheel/linux.py]: @joerick - I just opened https://github.com/joerick/cibuildwheel/pull/164 which has my original attempt with virtualenv alone.

**Comment by joerick** [comment]
Closing in favour of #164
